internal/workflow/activities/unarchive.go

```diff
@@ -0,0 +1,178 @@
+	}
+	if u == nil {
+		// Couldn't find an unarchiver, so this is probably a regular file.
+		// Return the source path unaltered, and IsDir false.
+		a.logger.V(2).Info("Unarchive: not an archive, skipping")
+		return &UnarchiveResult{DestPath: params.SourcePath}, nil
+	}
+
+	dest := filepath.Join(filepath.Dir(params.SourcePath), "extract")
+	if err := u.Unarchive(params.SourcePath, dest); err != nil {
+		return nil, fmt.Errorf("unarchive: unarchive: %v", err)
+	}
+
+	if params.StripTopLevelDir {
+		if err = stripDirContainer(dest); err != nil {
+			return nil, fmt.Errorf("unarchive: strip top-level dir: %v", err)
+		}
+	}
+
+	if err := fsutil.SetFileModes(dest, ModeDir, ModeFile); err != nil {
+		return nil, fmt.Errorf("unarchive: %v", err)
+	}
+
+	if err := os.Remove(params.SourcePath); err != nil {
+		a.logger.V(1).Info("Unarchive: couldn't delete source archive: %v", err)
+	}
+
+	return &UnarchiveResult{DestPath: dest, IsDir: true}, err
+}
+
+// Unarchiver returns the unarchiver suited for the archival format.
+func unarchiver(filename string) (archiver.Unarchiver, error) {
+	if iface, err := archiver.ByExtension(filename); err == nil {
+		if u, ok := iface.(archiver.Unarchiver); ok {
+			return u, nil
+		}
+	}
+
+	file, err := os.Open(filepath.Clean(filename))
+	if err != nil {
+		return nil, fmt.Errorf("open file: %v", err)
+	}
+	defer file.Close()
+
+	if u, err := archiver.ByHeader(file); err == nil {
+		return u, nil
+	}
+
+	return nil, nil
+}
+
+// stripDirContainer strips the top-level directory of a transfer.
+func stripDirContainer(dir string) error {
+	tld, err := topLevelDir(dir)
+	if err != nil {
+		return fmt.Errorf("get top-level dir: %v", err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "enduro-")
+	if err != nil {
+		return fmt.Errorf("make temp dir: %v", err)
+	}
+	defer func() { _ = os.RemoveAll(tmpDir) }()
+
+	// Move the top-level directory and contents to tmpPath.
+	tmpPath := filepath.Join(tmpDir, tld)
+	if err := os.Rename(filepath.Join(dir, tld), tmpPath); err != nil {
+		return fmt.Errorf("move: %v", err)
+	}
+
+	// Move the TLD contents back to the original path.
+	err = filepath.WalkDir(tmpPath, func(path string, d fs.DirEntry, err error) error {
+		if path == tmpPath {
+			return nil
+		}
+
+		if err := os.Rename(path, filepath.Join(dir, d.Name())); err != nil {
+			return fmt.Errorf("move to temp dir: %v", err)
+		}
+
+		// Don't descend into sub-directories.
+		if d.IsDir() {
+			return filepath.SkipDir
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("move back to top-level dir: %v", err)
+	}
+
+	return nil
+}
+
+// topLevelDir reads the directory at path and returns the name of it's
+// immediate sub-directory. If path contains anything other than a single
+// sub-directory then topLevelDirectory will return an error.
+func topLevelDir(path string) (string, error) {
+	r, err := os.Open(filepath.Clean(path))
+	if err != nil {
+		return "", fmt.Errorf("cannot open path: %v", err)
```

Comment:
os open will already tell us that we could not open a path. This is another case I think it might make sense to give more useful error messages or just return the base error.

Reply:
This error message was copied when I moved this code from `activities/bundle.go` - the only change I made was to remove the "errPrefix" string.

Reply:
I've updated this error message and the others in `topLevelDir()`.